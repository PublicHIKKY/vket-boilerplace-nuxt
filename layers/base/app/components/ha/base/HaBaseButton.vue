<template>
  <button
    class="ha-base-button"
    :disabled="disabled"
    :type="type"
    @click="onClick"
  >
    <slot>button label</slot>
  </button>
</template>

<script setup lang="ts">
export type ButtonType = 'button' | 'reset' | 'submit'

const props = withDefaults(
  defineProps<{
    disabled?: boolean
    type?: ButtonType
  }>(),
  {
    disabled: false,
    type: 'button',
  },
)
const emit = defineEmits<{
  (e: 'click', _e: MouseEvent): void
}>()
const onClick = (event: MouseEvent): void => {
  if (props.disabled) return
  emit('click', event)
}
</script>

<style lang="scss" scoped>
.ha-base-button {
  max-height: 100%;
  max-width: 100%;
}
</style>
