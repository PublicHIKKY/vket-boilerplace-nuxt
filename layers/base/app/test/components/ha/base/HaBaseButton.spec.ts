import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import HaBaseButton from '#base/app/components/ha/base/HaBaseButton.vue'

test('ref component', () => {
  expect(HaBaseButton).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HaBaseButton, {
    props: {
      type: 'button',
    },
  })
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

describe('props', () => {
  describe(':disabled', () => {
    // disabledは論理属性なので存在しているかどうかを確認する
    test('default is false', () => {
      const wrapper = mount(HaBaseButton)
      expect(wrapper.get('button').attributes('disabled')).toBeUndefined()
    })
    test('is disabled', () => {
      const wrapper = mount(HaBaseButton, { props: { disabled: true } })
      expect(wrapper.get('button').attributes('disabled')).toBeDefined()
    })
  })
  describe(':type', () => {
    test('default is button', () => {
      const wrapper = mount(HaBaseButton)
      expect(wrapper.get('button').attributes('type')).toBe('button')
    })
    test('pass prop', () => {
      const wrapper = mount(HaBaseButton, { props: { type: 'submit' } })
      expect(wrapper.get('button').attributes('type')).toBe('submit')
    })
  })
})

describe('emits', () => {
  test('click emits clickEvent', async () => {
    const wrapper = mount(HaBaseButton)
    await wrapper.get('button').trigger('click')
    expect(wrapper.emitted().click?.length).toBe(1)
  })
  test('click emits no event when disabled', async () => {
    const wrapper = mount(HaBaseButton, { props: { disabled: true } })
    await wrapper.get('button').trigger('click')
    expect(wrapper.emitted().click).toBeUndefined()
  })
})
