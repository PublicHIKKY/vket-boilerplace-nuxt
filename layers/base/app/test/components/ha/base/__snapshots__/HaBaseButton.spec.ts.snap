// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `"<button data-v-0844060a="" class="ha-base-button" type="button">button label</button>"`;
